### ISSUE TITLE
DOC: Add a doc page detailing advantages (and disadvantages?) vs. TF skelarn wrappers

### ISSUE THREAD

**Issue by adriangb** [adriangb/scikeras#132]
A basic question that users might ask before even trying SciKeras or reading the docs is: _why should I use SciKeras over tf.keras.wrappers.scikit_learn_?

To answer this, I think we should put 5-6 bullet points on our README.md as well as docs.

**Comment by adriangb** [comment]
Some ideas:
* SciKeras is actively maintained; the tensorflow wrappers [may be deprecated](https://github.com/tensorflow/tensorflow/pull/37201#pullrequestreview-391650001).
* Detailed [documentation](https://scikeras.readthedocs.io/en/latest/index.html).
* Full compatibility with the Scikit-Learn API, including grid searches, ensembles, transformers, etc.
* Support for pre-trained models.
* Support for dynamic input/output sizes.
* Support for hyperparameter tuning for estimators and losses.
* Support for multi-input and multi-ouput Keras models.
* Functional `random_state` for reproducible training.

**Comment by stsievert** [comment]
What downsides are there to using SciKeras? Does distributed training work with SciKeras?

I ask because I saw someone (@2over12?) looking to parallelize Keras training with Dask. They report that they...

> use[d] the keras scipy wrapper and a backing distributed optimization strategy [MirroredStrategy][1]] then use incremental training on it and it just works
>
> ... [s]o essentially what is happening as I understand it now [is that] the keras model fit is transparently splitting up a batch to each [model] replica and then syncing up gradients

This required both a TF cluster and a Dask cluster on the same set of machines. I think an example would be really useful to see.

[1]:https://www.tensorflow.org/api_docs/python/tf/distribute/MirroredStrategy

**Comment by adriangb** [comment]
This comparison was, specifically, to the old wrappers. I think a comparison to using Pure Keras (i.e. where does SciKeras break down) is warranted but can be done separately.

Regarding distributed training, I do not have much experience with it. Out of context, I do not really understand the above quote/comment, can you link to the full discussion? It sounds like an interesting use. If you would like to write something up regarding distributed training, I am more than happy to accept the contribution.

**Comment by 2over12** [comment]
So I guess this is in regards to the dask training with multiple gpus. I've reproduce the mnist scikeras stuff from here https://ml.dask.org/keras.html with local mirroring two the devices gpus. I believe it should be possible to also do this with https://www.tensorflow.org/api_docs/python/tf/distribute/experimental/MultiWorkerMirroredStrategy to run the model across multiple workers. It's pretty nice since with scikeras we can integrate directly into the dask preprocessing pipelines and take advantage of incremental hyper-parameter search  as well as sklearn stuff.

It's also worth noting the current page on dask+scikeras has an incorrect loss function https://www.tensorflow.org/tutorials/distribute/keras but that issue probably needs to be opened on the dask side.

```py
import tensorflow as tf
from tensorflow.keras.layers import Dense
from tensorflow.keras.models import Sequential
from scikeras.wrappers import KerasClassifier
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical
import numpy as np
from typing import Tuple
import dask_ml
from dask import array
import dask
#reference: https://www.tensorflow.org/tutorials/distribute/keras

def build_model(lr=0.01, momentum=0.9):
    # create a default mirror strategy which will have a device per gpu
    strat = tf.distribute.MirroredStrategy()
    with strat.scope():
        layers = [Dense(512, input_shape=(784,), activation="relu"),
                Dense(10, input_shape=(512,), activation="softmax")]
        model = Sequential(layers)

        opt = tf.keras.optimizers.SGD(
            learning_rate=lr, momentum=momentum, nesterov=True,
        )
        model.compile(loss="sparse_categorical_crossentropy", optimizer=opt, metrics=['accuracy'])
    return model


def get_mnist() -> Tuple[np.ndarray, np.ndarray]:
    (X_train, y_train), _ = mnist.load_data()
    X_train = X_train.reshape(X_train.shape[0], 784)
    X_train = X_train.astype("float32")
    X_train /= 255
    return X_train, y_train

if __name__ == "__main__":
    dask.distributed.Client()
    
    X,y = get_mnist()
    model = KerasClassifier(build_fn=build_model, lr=0.1, momentum=0.9)
    inc_mod = dask_ml.wrappers.Incremental(model)
    X=array.from_array(X,chunks=10000)
    y=array.from_array(y,chunks=10000)



    inc_mod.fit(X,y)
```

**Comment by adriangb** [comment]
To be honest, I'm not sure what the differences/advantages to TF's distributed functionalities might be over Dask's. But I think this may be a good example of why it's important to build an ecosystem with consistent APIs: you get solutions to complex problems like this _for free_.

**Comment by stsievert** [comment]
Woah https://github.com/adriangb/scikeras/issues/132#issuecomment-735031792 is interesting. What's the data flow exactly? If I'm reading it correctly, the following happens:

1. Dask creates (say) 32 models.
2. Every model is sent to (say) 4 GPUs
3. Dask controls which models get fit.
4. Fitting one model with batch size (say) 256:
    * Each worker computes 64 gradients.
    * There's a synchronous distributed/all-reduce parameter server (PS) between GPUs ([MultiWorkerMirroredStrategy][1])

Correct me if I'm wrong. I don't see any GPU-specific code to put the model on the GPU, but it looks like the distributed strategy takes care of that: https://keras.io/guides/distributed_training/

Also, [dask-cuda] might be useful if you want to limit the number of models per GPU.

> I'm not sure what the differences/advantages to TF's distributed functionalities might be over Dask's

It looks like they're operating independently in https://github.com/adriangb/scikeras/issues/132#issuecomment-735031792. From Dask's point of view, `model` is on the CPU and calling `model.fit` is a normal CPU operation. However, Keras handles putting `model.fit` on the GPU and the parallelism (with the data flow above). Is that accurate @2over12?



[1]:https://www.tensorflow.org/api_docs/python/tf/distribute/experimental/MultiWorkerMirroredStrategy

**Comment by 2over12** [comment]
@stsievert  Yeah so not quite. Dask Incremental executes a single model in serial. Each fitting task on each block of the array is intentionally dependent on the previous model fit. Dask would have no way to merge separate models if we executed many models in parallel (you could group these models into an ensemble with something like `ensemble.BlockwiseVotingClassifier` but generally I'd prefer a single model that has seen all the data). So what's happening here is dask workers can be used to create the dask array and preform transformations in parallel. In our use-case we use these workers to read in a dataframe and transform it into the x and y array as well as do some last minute transformations before training. Then each serial fit task gets fed chunks as they are available. So essentially it is a many to 1 relationship where the are many workers generating and manipulating chunks and 1 worker running the model. The distributed strategy then lets us explode out each of these serial training tasks into parallelized training where batches are split up between the gpus. So as you said if you have 4 GPUs and a 256 batch size each GPU runs 64 samples on the model and computes gradients which are merged to do the final updates. I believe you could replace tf.distribute.MirroredStrategy() with the multiworker strategy in order to explode out the batch fitting to a multinode setup of gpus. 

It's worth noting the mirrored strategies in tensorflow are generally going to be a bit slower than something like horovod https://github.com/horovod/horovod where models do not require synchronization and instead the average of each worker's gradient is applied to each variable. Horovod however would be annoying to integrate with dask due to how it spins up it's own clusters etc. Ideally we would have the capability to run some distributed SGD in dask like as done in spark and here https://zhangyuc.github.io/splash/ but tensorflow mirrored strategies are a decent start
